### PR TITLE
Make SecHub&PDS cancel cluster ready, provide config defaults and events

### LINKED ISSUE #1462: PDS config file parameters can have default settings
## Situation
Currently we have
- mandatory
- optional
parameter definitions in our PDS config files.

Here we define which parameters are accepted by the dedicated products.
But we must define this always at SecHub side, even when there is a good default value.

## Wanted
Inside the PDS config file it shall be possible to define default values for product

## Solution
- Enhance `PDSProductParameterDefinition` to have also a default value
- when we inspect/fetch the parameters and the parameter is empty we will use the default value (if existing)

### LINKED ISSUE #1463: Introduce PDS workspace `events` and provide `cancel` event handling for PDS scripts and wrappers
## Situation
When we have a PDS which communicates with a product which has an additional job management - e.g. Checkmarx (see #1415 for details ), we currently can cancel the PDS job, but this will only terminate the PDS script process. The external job is still running.

## Wanted
A general solution for PDS cancel handling for scripting parts.

## Solution
- in general: 
  -  workspace service will be able to send execution events to workspace folder `$workspaceFolder/events/`
  - inside this folder there will by JSON data of a simple java.util.map (key,value) 
  - the map will contain at least "vent.created" as UTC timestamp in ISO 8601
  - PDS cancel operation shall write a `$workspaceFolder/events/cancel_requested.json` file
- the cancel operation on PDS side must be changed as well:
  - introduce new PDS parameter `sechub.pds.config.execute.event.cancelrequest.seconds.waitforscript`. 
    - The falback default value in code is always : 0
    - With #1462 implemented a PDS-solution can now define this directly (and override 0) in its PDS config file, so 
      admins do not need to setup this parameter all the time, but they CAN if wanted.
  - process may not be stopped directly. But at least time for cancel wait will be accepted

### LINKED ISSUE #1474: Existing cancel operation on `PDS` does not terminate the PDS launcher script process
## Situation
When executing the "cancel job" operation via REST the job is marked as canceled in DB, but the process is still running

## Wanted
A cancel shall terminate the process

## Solution
Change `PDSExecutionCallable`

### LINKED ISSUE #1532: Rename "stop" parts inside adapters to "cancel"
## Situation

Inside adapter we have the term "stop". But everywhere else we talk about "cancel".

## Wanted
In adapters it shall also be "cancel" - so it is clear that we are talking about the same action

## Solution
Rename

### LINKED ISSUE #1547: A cancel operation shall end with execution state `CANCELED` instead of `CANCEL_REQUESTED`
## Situation
Currently a cancel operation for a sechub job, does end up with `CANCEL_REQUESTED` as execution state.
This is a little bit confusing because the question comes up : "... hm... okay, its requested... but it is really done?"

## Wanted
A cancel operation shall end with execution state `CANCELED` instead of `CANCEL_REQUESTED`

## Solution
- Introduce a new value inside `ExecutionState` enum: `CANCELED`
- Check for tests and logic where `CANCELED_REQUESTED` is checked - and add `CANCELED` to this checks as well 
  (to avoid side effects)
- Listen to events having message id `JOB_CANCELED` inside scheduler domain and change the job execution state in this case to `CANCELED`

### LINKED ISSUE #1548: Introduce new message id `JOB_CANCELLATION_RUNNING`
## Situation
With #1547 we  want an execution state extension, so we have then "CANCELED" as final state for cancel
But the current event message ID  which is send for the CANCEL_REQUESTED execution state is "CANCELED"... So what do we use for the "final cancel" event...

## Wanted
Clear naming for the current cancel state

## Solution
- We rename existing message id enum `JOB_CANCELED` to `JOB_CANCELLATION_RUNNING` inside IDE which does rename all former usages automatically
- We introde a "new" enum with message id `JOB_CANCELED` and use this for
- #1547  and also for
- #1549

### LINKED ISSUE #1549: Implement missing scan cancelation inside existing Job cancel operation
## Situation
At the moment, SecHub does only cancel its own job, but does not call any adapter to cancel the adapter operation!

## Wanted
When the job has been canceled, the adapters shall be triggered again and handle the cancel in their own way (if implemented in adapters)

## Solution
- Change event handling - needs #1547  and #1548  to be implemented - when the step of scan cancelation is done, this will trigger the final   scan done event (then the "new" `JOB_CANCELED` )
- Provide a simple queue mechanism for the execution of the cancel operations (we do not want to block the event queue!)
- Ensure the cancel operation does exactly trigger the same product executors to do the `cancel` operation like done for the former `start scan` operation!

### LINKED ISSUE #1550: Fix typo in spelling of cancellation
## Situation
Cancellation was often wrong written by using only one "l" instead of two.

## Wanted
All text parts shall contain "cancellation" correctly

## Solution
Fix it inside sources

### LINKED ISSUE #1554: ScanJobExecutor cannot fetch real cancelation state
## Situation
After analyzing the current scan implementation for 
- #1474 and also for 
- #1549 

It became clear that there is an extreme problem for the cancelation process at the moment:

`ScanJobExecutor` which is the main class for the scan operation does use the `ScanProgressMonitor` to check if the scan process has been canceled. But the monitor does check for the spring batch status .

On the other hand, the `SchedulerCancelJobService` try to cancel when domain the event "CANCEL REQUESTED" comes up and tries to stop the Batch processing via `SchedulerCancelBatchJobService` - but this does not work.

If it would work, the complete `ScanJobExecutor` execution would be terminated/ the cancel operations would never work!

## Wanted
Change to more stable variant which works!

## Solution
The messaging and the logic shall now no longer try to stop the spring batch operation!

Instead  the job status data shall be retrieved directly by the scheduler job status (also by synchronous events to scheduler domain, but no longer with spring batch data).

When status is CANCEL_REQUESTED, the cancel mechanism shall start.

### LINKED ISSUE #1555: Provide possibilty to turn off verbose logging of PDS heartbeats
## Situation
Sometimes it is cumbesome to have always the heartbeat logging events inside our logs - e.g. when debugging a PDS instance because of a new feature. The hearbeat can be disabled, but when we just want to get rid off the logging there is no way.

## Wanted
The default behaviour shall be as before - heartbeat logging shall be done on info log level.
But there shall be a possibility to turn it off.

## Solution
We introduce new spring boot value:
`sechub.pds.config.heartbeat.verbose.logging.enabled` - the default is `true`

### LINKED ISSUE #1556: Log errors fetched by PDS error controller
## Situation
Currently every exception which is handled by `PDSServerErrorController` does not show up inside console.
It is very cumbersome to debug, develop or to maintain PDS instances when such errors/problems are just swallowed

## Wanted
Those exceptions shall be logged as well

## Solution
Add logging mechanism inside `PDSServerErrorController`:
- introduce `sechub.pds.server.errorcontroller.log.errors` (default:true)
- when enabled the exception shall be resolved by injected error attributes and logged, together with information about the data send back to server (e.g. HTTP status, if stacktrace has been send etc.)

### LINKED ISSUE #1642: Parallelize normal unit tests in builds
### Situation
The build time for the "normal" java build takes a long time (currently ~6-7 minutes at CI build with GitHub actions)

### Wanted
Unit tests should be run parallel where possible

### Solution
We add `--parallel` to the java gradle build parts (jenkins + github workflows) - but NOT for integration tests!

### LINKED ISSUE #1652: Drop Abandonable, SecHubExecutionAbandonedException and related
## Situation
With the changes done in #1474 the spring batch operation to "abandon" is no longer used and no class is implementing 
the interface `Abandonable` any longer and this means that `SecHubExecutionAbandonedException` can never be thrown

We also plan to remove the complete spring batch framework usage in near future - so we can use this for a "pre-cleanup". 

## Wanted
Clean code without unnecessary parts.

## Solution
Remove interface, exception and usages

### PR DESCRIPTION
**Pull request by de-jcup**
This PR
- closes #1474 
- closes #1462
- closes #1463 
- closes #1532 
- closes #1547 
- closes #1548
- closes #1549
- closes #1550
- closes #1554
- closes #1555
- closes #1556
- closes #1642
- closes #1652

### CODE REVIEW

**Review by winzj — CHANGES_REQUESTED** [review]
Looks very good to me and I'm looking forward to some of the new features and improvements. Thank you for the explanations alongside the review.

There are some small changes that might need to be made. Please have a look at the comments for details.

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/ScanJobExecutor.java [hunk 1]
```diff
@@ -111,63 +127,60 @@ private void handleErrors(SecHubExecutionException exception) throws SecHubExecu
         throw exception;
     }
 
-    private void handleCanceled(CanceableScanJobRunnable canceableJobRunner, UUID sechubJobUUID) throws SecHubExecutionAbandonedException {
+    private void handleCancelRequested(ScanJobExecutionRunnable executionRunable, UUID sechubJobUUID) throws SecHubExecutionAbandonedException {
         LOG.info("Received cancel signal, so start canceling job: {}", sechubJobUUID);
-        canceableJobRunner.cancelScanJob();
-        if (!(progress instanceof Abandonable)) {
+        cancelExecutoinRunnableAndFailIfAbandoned(executionRunable, sechubJobUUID);
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/ScanJobExecutor.java:132]: The interface `Abandonable` has no longer any implementation (was introduced for batch jobs).
So I will remove the interface and all implementations instead (applying YAGNI pattern..)

[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/ScanJobExecutor.java:132]: Created #1652 for this  and apply after  review changes commit  in another one.

**Review by winzj — APPROVED** [review]
Good job, looks good

**Pull request merged**